### PR TITLE
Feat: implemented withdrawal_utils functions

### PR DESCRIPTION
**Pull request by ametel01**
# Pull Request type

- Feature

Resolves: #359 

## Other information
Implemented using IStrictBank interface instead of the IWithdrawalVault as found in the Solidity version. Please confirm if this needs to be updated.

### CODE REVIEW

File: src/withdrawal/withdrawal_utils.cairo [hunk 1]
```diff
@@ -187,14 +536,110 @@ fn swap(
     receiver: ContractAddress,
     ui_fee_receiver: ContractAddress,
     should_unwrap_native_token: bool,
-) -> (ContractAddress, u128) { // TODO
-    (0.try_into().unwrap(), 0)
+) -> (ContractAddress, u128) {
+    let mut cache: SwapCache = Default::default();
```
Comment (sparqet) [inline src/withdrawal/withdrawal_utils.cairo:540]: Declare the struct with angle brackets

**Review by sparqet — CHANGES_REQUESTED** [review]

File: src/withdrawal/withdrawal_utils.cairo [hunk 2]
```diff
@@ -108,18 +146,133 @@ fn create_withdrawal(
...
+
+    params.execution_fee = wnt_amount.into();
+
+    market_utils::validate_enabled_market(data_store, params.market);
+
+    market_utils::validate_swap_path(data_store, params.long_token_swap_path);
+
+    market_utils::validate_swap_path(data_store, params.short_token_swap_path);
+
+    let mut withdrawal: Withdrawal = Default::default();
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/withdrawal/withdrawal_utils.cairo:179]: done

File: src/withdrawal/withdrawal_utils.cairo (hunk 1 above)
Comment (ametel01) [reply src/withdrawal/withdrawal_utils.cairo:540]: implemented

File: src/withdrawal/withdrawal_utils.cairo [hunk 3]
```diff
@@ -135,13 +288,49 @@ fn execute_withdrawal(params: ExecuteWithdrawalParams) { // TODO
...
+    data_store.remove_withdrawal(key, withdrawal.account);
+
+    withdrawal_vault
+        .transfer_out(
+            withdrawal.market, withdrawal.account, withdrawal.market_token_amount
+        ); //, false);
+
+    event_emitter.emit_withdrawal_cancelled(key, reason, reason_bytes.span());
+
+    let mut event_data: EventLogData = Default::default();
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/withdrawal/withdrawal_utils.cairo:321]: Thanks for your review. I've eliminated all references to EventLogData and have only left the event emitter call in place. Could you clarify the point you were making earlier?

File: src/event/event_utils.cairo [hunk 4]
```diff
@@ -1,5 +1,163 @@
-#[derive(Drop, starknet::Store, Serde)]
+// NOTE: temporary mock in order to complet withdrawal_utils.
+use starknet::ContractAddress;
+
+#[derive(Default, Drop, Serde)]
 struct EventLogData {
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/event/event_utils.cairo:5]: Thanks for your review. I've eliminated all references to EventLogData and have only left the event emitter call in place. Could you clarify the point you were making earlier?

File: src/event/event_utils.cairo [hunk 5]
```diff
@@ -1,5 +1,163 @@
...
+    int_items: IntItems,
+    bool_items: BoolItems,
+    bytes32_items: Bytes32Items,
+    bytes_items: BytesItems,
+    string_items: StringItems,
 }
+
+
+#[derive(Default, Drop, Serde)]
+struct AddressItems {
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/event/event_utils.cairo:17]: removed

File: src/event/event_utils.cairo [hunk 6]
```diff
@@ -1,5 +1,163 @@
...
+}
+
+#[derive(Default, Drop, Serde)]
+struct StringArrayKeyValue {
+    key: felt252,
+    value: Array<felt252>
+}
+
+trait EventUtilsTrait<T, A> {
+    fn set_item(ref self: T, key: felt252, value: A);
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/event/event_utils.cairo:149]: removed

File: src/withdrawal/withdrawal_utils.cairo [hunk 7]
```diff
@@ -108,18 +146,133 @@ fn create_withdrawal(
...
+        );
+    }
+
+    let result = execute_withdrawal_(@params, withdrawal);
+
+    params.event_emitter.emit_withdrawal_executed(params.key);
+
+    let mut event_data: EventLogData = Default::default();
+
+    event_data.address_items.set_item('output_token', result.output_token);
```
[the remark this answers is not in the document]
Comment (ametel01) [reply src/withdrawal/withdrawal_utils.cairo:255]: Thanks for your review. I've eliminated all references to EventLogData and have only left the event emitter call in place. Could you clarify the point you were making earlier?

[the remark this answers is not in the document]
Comment (sparqet) [reply src/withdrawal/withdrawal_utils.cairo]: In our implementation, each event is emitted by a specific emitter

**Review by sparqet — APPROVED** [review]
LGTM

**Pull request merged**